# Agenda
## Basic Operations with OpenCV:

1. Importing OpenCV
2. Loading an image
3. Image Shape/ Resolution
4. Displaying the image
5. Converting to Grayscale
6. Converting image into individual color channels (Red, Green, Blue)

In [3]:
# Press SHIFT +  ENTER to runa a particualr cell. Control will move to next cell.
# Press CTRL + ENTER to run a particular cell. Control will remain on that cell.
# When a cell is executing, there is an * between the [ ] on the left
# Press CTRL + ENTER to run this line. Control will remain on same cell

## Image Processing Using OpenCV

### What is OpenCV.

OpenCV (Open Source Computer Vision Library) is an open source computer vision and machine learning software library. OpenCV was built to provide a common infrastructure for computer vision applications and to accelerate the use of machine perception in the commercial products. 

Doc: https://opencv.org/about/

### Let's start by importing the OpenCV libary 

In [4]:
# This command will add opencv in your current notebook environment.
# ! is deliberate, this will install open-cv in your current notebook environment.
!pip install opencv-python

You should consider upgrading via the '/home/ubuntu/notebook_2/bin/python -m pip install --upgrade pip' command.


In [5]:
# OpenCV takes a couple seconds to import the first time

import cv2

# Question: What does the import cv2 statement do?

1. Imports the SciPy library for numerical processing.

2. Imports the NumPy library for numerical processing.

3. Imports our OpenCV Python bindings.

4. Displays an image to our screen.

In [6]:
# Now let's import numpy
# We use as np, so that everything we call on numpy, we can type np instead
# It's short and looks neater

import numpy as np 

## Reading, writing and displaying images with OpenCV
## Let's now load our first image


#### imread() = https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_gui/py_image_display/py_image_display.html#read-an-image

#### imshow() = https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_gui/py_image_display/py_image_display.html#display-an-image
 

In [7]:
# Load an image using 'imread' specifying the path to image
input = cv2.imread('./images/input.jpg')

# To display our image variable, we use 'imshow'
# The first parameter will be title shown on image window
# The second parameter is the image varialbe
cv2.imshow('Input Image', input)


# cv2.waitKey() is a keyboard binding function. 
# Its argument is the time in milliseconds. 
# The function waits for specified milliseconds for any keyboard event. 
# If you press any key in that time, the program continues. 
# If 0 is passed, it waits indefinitely for a key stroke.
cv2.waitKey(0) 

# This closes all open windows 
# Failure to place this will cause your program to hang
cv2.destroyAllWindows()

In [9]:
# Same as above without the extraneous comments

import cv2 

input = cv2.imread('./images/input.jpg')

cv2.imshow('Input Image', input)
cv2.waitKey(0)
cv2.destroyAllWindows()

### Let's take a closer look at how images are stored

In [10]:
# Import numpy
import numpy as np

In [11]:
print(input.shape)
# The shape of an image is accessed by img.shape. 
# It returns a tuple of the number of rows, columns, and channels (if the image is color):

(830, 1245, 3)


#### Shape gives the dimensions of the image array

The 2D dimensions are 830 pixels in high bv 1245 pixels wide.
The '3L' means that there are 3 other components (RGB) that make up this image.


By shape of the image, we mean the shape of the NumPy array. As you see from executing the code, the matrix consists of 830 rows and 1245 columns.

The last value 3 indicates the no. of channels.

3 = coloured image
1 = black and white/grayscale image

In [12]:
# Let's print each dimension of the image

print ('Height of Image:', int(input.shape[0]), 'pixels')
print ('Width of Image: ', int(input.shape[1]), 'pixels')

Height of Image: 830 pixels
Width of Image:  1245 pixels



# Question: Given the following NumPy array shape, how would we interpret the width, height, and number of channels in the image: (400, 600, 3):

1. Width=600, height=400, channels=3

2. Width=600, height=3, channels=400

3. Width=400, height=600, channels=3

4. width=3, width=600, channels=400

### How do we save images we edit in OpenCV?

#### imwrite() = https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_gui/py_image_display/py_image_display.html#write-an-image

In [13]:
# Simply use 'imwrite' specificing the file name and the image to be saved
# imwtite() returns True if image saved successfully
# we will use our previous image to save with name "imwrite"
cv2.imwrite('imwrite.jpg', input)
cv2.imwrite('imwrite.png', input)

True

## Question:  Save image in png format. 

In [15]:
# name.png
cv2.imwrite('......', input)  

True

# Color Space and Color Mapping
## Grayscaling

#### Grayscaling is process by which an image is converted from a full color to shades of grey (black & white)

### Let convert our color image to grayscale

By default OpenCv reads image in BGR color space.

In [16]:
# This cell will show orignal image
import cv2
import numpy as np

# Load our input image
image = cv2.imread('./images/input.jpg')
cv2.imshow('Original Input', image)
cv2.waitKey(0)
cv2.destroyAllWindows()

#### cvtColor = https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_imgproc/py_colorspaces/py_colorspaces.html

In [17]:
# This cell will convert orignal image into gray image
# We use cvtColor, to convert to grayscale
gray_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
# For BGR -> Gray conversion we use the flags cv2.COLOR_BGR2GRAY.

cv2.imshow('Grayscale', gray_image)
cv2.waitKey(0)
cv2.destroyAllWindows()

## Let's take a closer look at color spaces

You may have remembered we talked about images being stored in RGB (Red Green Blue) color Spaces. Let's take a look at that in OpenCV.

Let's look at the image shape again. The '3L' 

In [18]:
import cv2
import numpy as np

image = cv2.imread('./images/input.jpg')

### Let's look at the individual color levels for the first pixel (0,0)

In [19]:
# BGR Values for the first 0,0 pixel
B, G, R = image[10, 50] 
print (B, G, R)
print (image.shape) 
# The shape of an image is accessed by img.shape. 
# It returns a tuple of the number of rows, columns, and channels (if the image is color)

13 19 32
(830, 1245, 3)


In [20]:
image[0, 0]

array([12, 18, 31], dtype=uint8)

In [21]:
cv2.imshow('Image', image)
cv2.waitKey(0)
cv2.destroyAllWindows()

# Question: OpenCV stores RGB pixels in what order?

1. GBR

2. RGB

3. BRG

4. BGR

### Let's see what happens when we convert it to grayscale

In [22]:
# to convert our original image from the BGR color space to gray, we use the code COLOR_BGR2GRAY.
gray_img = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
print (gray_img.shape)
print (gray_img[10, 50]) 

(830, 1245)
22


In [23]:
cv2.imshow('Gray Image', gray_img)
cv2.waitKey(0)
cv2.destroyAllWindows()

#### It's now only 2 dimensions. Each pixel coordinate has only one value (previously 3) with a range of 0 to 255

In [24]:
gray_img[0, 0]

21

## Splitting Color Images into Individual color
### As a color image consist of three color channels RGB (Red, Green, Blue) but in OpenCV format it is BGR

In [25]:
# Lets read a color image

color_image = cv2.imread('./images/beatle.jpg')
cv2.imshow('Color Image', color_image)
cv2.waitKey(0)
cv2.destroyAllWindows()

### Split Image into BGR channel

In [26]:
# Break it into BGR
b, g, r = cv2.split(color_image)

In [27]:
# blue
cv2.imshow('Blue Image', b)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [28]:
cv2.imshow('Green Image', g)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [29]:
cv2.imshow('Red Image', r)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [30]:
restore_image = cv2.merge((b,g,r))
cv2.imshow('Restored Image', restore_image)
cv2.waitKey(0)
cv2.destroyAllWindows()

## Display image in particular color channel

In [31]:
# setting green and red channel to zero to display only blue channel
blue_image = color_image.copy()
blue_image[:, :, 1] = 0
blue_image[:, :, 2] = 0
cv2.imshow('Blue Image', blue_image)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [32]:
# setting blue and red channel to zero to display only green channel
green_image = color_image.copy()
green_image[:, :, 0] = 0
green_image[:, :, 2] = 0
cv2.imshow('Blue Image', green_image)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [33]:
# setting blue and red channel to zero to display only red channel
red_image = color_image.copy()
red_image[:, :, 0] = 0
red_image[:, :, 1] = 0
cv2.imshow('Red Image', red_image)
cv2.waitKey(0)
cv2.destroyAllWindows()

### Another useful color space is HSV
Infact HSV is very useful in color filtering.



In [34]:
#H: 0 - 180, S: 0 - 255, V: 0 - 255
import cv2
image = cv2.imread('./images/input.jpg')

hsv_image = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)

cv2.imshow('HSV image', hsv_image)
cv2.waitKey()
cv2.destroyAllWindows()

### To get image into HSV channels separately, we have to select channel 1 for Hue, 2 for Saturation and 3 for Vue

In [35]:
#  Hue Channel
cv2.imshow('Hue channel', hsv_image[:, :, 0])
cv2.waitKey()
cv2.destroyAllWindows()

In [36]:
# Saturation Channel
cv2.imshow('Saturation channel', hsv_image[:, :, 1])
cv2.waitKey()
cv2.destroyAllWindows()

## Question: Show image in Value Channel

In [42]:
#  Value Channel
cv2.imshow('Value channel', hsv_image[:, :, ???????])
cv2.waitKey()
cv2.destroyAllWindows()